### PR TITLE
TIP-1159: Add new calculator

### PR DESCRIPTION
**Pull request by pierallard**
[wip]

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Completeness/SqlGetCompletenessFamilyMasks.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    SELECT 
+        family.code AS familyCode,
+        channel.code AS channelCode,
+        JSON_ARRAYAGG(attribute.code) AS attributeCodes
+    FROM pim_catalog_attribute_requirement requirement
+    INNER JOIN pim_catalog_attribute attribute ON attribute.id=requirement.attribute_id
+    INNER JOIN pim_catalog_channel channel ON channel.id=requirement.channel_id
+    INNER JOIN pim_catalog_family family ON family.id=requirement.family_id
+    WHERE requirement.required=1
+        AND family.id IN (:familyCodes)
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Completeness/SqlGetCompletenessFamilyMasks.php]: did you mean `family.code` ?

File: src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessCalculator.php [hunk 2]
```diff
@@ -1,131 +1,66 @@
...
-                    $locale,
-                    $incompleteValues->attributes(),
-                    $incompleteValues->count(),
-                    $requiredValues->count()
-                );
-            }
+        $result = [];
+        foreach ($products as $product) {
+            $familyMask = $familyMasks[$product->familyCode()];
+            // TODO Need the activated locales for this family. Not sure if it depends of the channel.
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessCalculator.php]: it does depend on the channel :)

File: src/Akeneo/Pim/Enrichment/Component/Product/Completeness/SqlGetCompletenessFamilyMasks.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    /**
+     * @param string[] $familyCodes
+     *
+     * @return CompletenessFamilyMask[]
+     */
+    public function fromFamilyCodes(array $familyCodes): array
+    {
+        $sql = <<<SQL
+SELECT 
+    familyCode,
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Completeness/SqlGetCompletenessFamilyMasks.php]: you will  also need locale codes, and <all_channels> / <all_locales> for non-scopable / non localizable attributes

File: src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessFamilyMask.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    public function getCompletenessCollection(Product $product): ProductCompletenessCollection
+    {
+        $productMask = $product->getMask();
+
+        return new ProductCompletenessCollection($product->getId(), array_map(
+            function (CompletenessFamilyMaskPerChannelAndLocale $familyMaskPerChannelAndLocale) use ($productMask) {
+                $diff = array_diff($familyMaskPerChannelAndLocale->mask(), $productMask);
+
+                $missingAttributeCodes = array_map(function (string $mask) {
+                    return substr($mask, 0, strpos($mask, '-'));
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessFamilyMask.php]:
```suggestion
                    return strstr($mask, '-', true);
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Completeness/Product.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    public function __construct(
+        int $id,
+        string $identifier,
+        string $familyCode,
+        array $rawValues
+    ) {
+        $this->id = $id;
+        $this->identifier = $identifier;
+        $this->familyCode = $familyCode;
+        $this->rawValues = $rawValues;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Completeness/Product.php]: you don't need to store the raw values, you could build the mask at instantiation and store it instead

Comment (pierallard) [reply src/Akeneo/Pim/Enrichment/Component/Product/Completeness/Product.php]: Not sure of this. I prefer wait to know how to manage properly the specific cases (prices e.g.)

File: src/Akeneo/Pim/Enrichment/Component/Product/Completeness/Model/Product.php [hunk 6]
```diff
@@ -60,4 +64,39 @@ public function getId(): int
...
+     *
+     * @return string
+     */
+    private function formatAttributeCode(string $attributeCode, $value)
+    {
+        if (is_array($value)) {
+            $isPrice = true;
+            $currencies = [];
+            foreach ($value as $v) {
+                if (!(isset($v['amount']) && isset($v['currency']))) {
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Completeness/Model/Product.php]: check if $v is an array

**Pull request closed without merging**